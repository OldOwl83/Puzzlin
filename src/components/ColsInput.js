import { useContext, useState, useEffect } from "react";
import { PuzzleGrid } from "../RompeCocosApp";

export const ColsInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ input, setInput ] = useState( puzzleGrid[0] );
    
    useEffect(() => {
        
        if( input >= 3 || input == '' )
        {
            if( input >= 3 )
                setPuzzleGrid( [ Number( input ), puzzleGrid[1] ] );
            else
                setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
        }else
        {
            setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
            setInput( 3 );
        }
    }, [ input, setPuzzleGrid ]);

    return <input 
                type="number" 
                value={ input }
                min="3"
                onChange={ ( e ) => setInput( e.target.value) } 
            />
}
