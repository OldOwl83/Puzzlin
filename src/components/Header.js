import React from 'react';
import { Link, useLinkClickHandler } from 'react-router-dom';

export const Header = () => {

    const linker = useLinkClickHandler( '/', { 
        target: '_self',
        replace: false,
        state: {
            contScroll: 0,
            selectedPhoto: 'images/portada1.png',
            author: 'Mauro Donnantuoni',
        }
    });

    return (
        <>
        <Link to='/' id='toHomeLink' onClick={ linker }>
            <header>
                <h1>Puzzlin</h1>
                <div id="logoContainer">
                    <img src="images/rompecabezas.png" alt="logo" />
                </div>
            </header>
        </Link>

        <hr />
        </>
    );
};
