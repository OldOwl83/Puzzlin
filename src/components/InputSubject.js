import { useState } from "react";
import { PropTypes } from 'prop-types';
import { useNavigate } from "react-router-dom";

import '../styles/InputSubject.css';

export const InputSubject = ( { setSearch } ) => {

    // console.log( 'InputSubject' );

    const navigate = useNavigate();

    const [ input, setInput ] = useState( '' );
    
    const submitInput = (e) => {
        e.preventDefault();
        
        setInput( '' );

        setSearch( input );

        navigate( `../${ input }` );
    }

    return (

        <h2>Input a subject and select a photo. Then, press on the preview to start.
            <form onSubmit={ submitInput }>
                <input 
                    id="inputSubject"
                    type="text"
                    autoComplete="on"
                    value={ input }
                    onChange={ (e) => setInput( e.target.value ) }
                />
            </form>
        </h2>
    )
}

InputSubject.propTypes = {

    setSearch: PropTypes.func.isRequired,
}