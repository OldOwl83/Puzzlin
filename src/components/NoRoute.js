import React from 'react';

export const NoRoute = () => {
  return <p className='message'>Nonexistent path.</p>;
};
