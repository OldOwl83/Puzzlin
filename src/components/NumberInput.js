/*
* General purpose number input. It receives by parameters a function which changes an external value or state, the minimum and maximum numbers allowed and the initial value. If the user tries to exceed the set range, it automatically sets the minimum or maximum accordingly. It sets the value as an integer data type.
*/

import { useState, useEffect, memo } from "react";

export const NumberInput = memo(( { setValue, min, max, initialValue }) => {

    const [ input, setInput ] = useState( initialValue );
    
    useEffect(() => {
    
        if( ( input >= min && input <= max ) || input === '' ) //It must include '' to allow erase manually the text in the input box.
        {
            if( input >= min && input <= max )
                setValue( Number( input ) );
            else 
                setValue( Number( min ) ); // If the input box remain empty, the set value is the minimum.
        }else if( input > max )
        {
            setValue( Number( max ) );
            setInput( max );
        }else
        {
            setValue( Number( min ) )
            setInput( min );
        }
    }, [ input, min, max, setValue ]);

    return (
    
        <input 
            type="number" 
            value={ input }
            min={ min }
            max={ max }
            onChange={ ( e ) => setInput( e.target.value ) } 
        />
    );
});
