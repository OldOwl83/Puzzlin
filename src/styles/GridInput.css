#GridInput{
    display: flex;
    justify-content: space-between;
    margin: 10px auto;
    width: 90%;
}

#GridInput p{
    width: 36%;
    padding: 10px 0;

    text-align: center;
    font-family: serif;
    color: rgb(1, 80, 83);
    font-size: 1.6rem;
    font-weight: bolder;
    
    background-color: rgb(230, 185, 127);
    border: 4px solid beige;
    border-radius: 10px;
}

#GridInput input{
    width: 20%;
    margin-left: 14px;
    padding: 6px;

    font-size: 1.4rem;
    font-family: 'Cantarell', sans-serif;
    font-weight: bold;
    text-align: center;

    /* border: 4px solid rgb(230, 185, 127); */
    border-radius: 6px;
    background-color: beige;
}