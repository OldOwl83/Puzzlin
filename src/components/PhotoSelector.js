import { useState, useEffect, useContext } from "react";

import { InputSubject } from "./InputSubject";
import { Preview } from './Preview';
import { useFetchPhotos } from "../hooks/useFetchPhotos";
import { PhotoShowcase } from "./PhotoShowcase";

import { MainPhotoContext } from "../RompeCocosApp";

import '../styles/PhotoSelector.css';
import 'animate.css';

export const PhotoSelector = () => {

    console.log('PhotoSelector');

    const [search, setSearch] = useState( '' );

    const { setPhotoPuzzle } = useContext( MainPhotoContext );

    // const [ imgSelected, setImgSelected] = useState( "images/tapa.png" );

    const { photos, loading } = useFetchPhotos( search );

    useEffect(() => {
        if(photos.length > 0)
            setPhotoPuzzle( photos[0].urlRegular );
    }, [ photos ])

    return (
        <>
            <InputSubject 
                setSearch={ setSearch } 
             />

            <div id="selectorContainer">
                
                <PhotoShowcase photos={ photos } />

                { loading && <p className="message">Loading...</p> }
                    
                { !photos && <p className="message">Connection to images provider failed.</p> }

                { search !== '' && !loading && photos.length < 1 && <p className="message">There were no matches with the search words.</p> }

                { (search === '' || photos.length > 0) && !loading && <Preview /> }
            </div>
        </>
    )
}