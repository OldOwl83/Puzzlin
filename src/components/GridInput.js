/*
* It handles the contextual values of the puzzle grid. These values represent the columns and rows to cut out the puzzle photo and are in an array with two indices, which is a state belonging to the <Puzzlin /> component.
*/

import { useContext, useState, useEffect } from "react";

import { PuzzleGrid } from "../Puzzlin";
import { NumberInput } from "./NumberInput";

import "../styles/GridInput.css";

export const GridInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ cols, setCols ] = useState( puzzleGrid[0] );
    const [ rows, setRows ] = useState( puzzleGrid[1] );
    
    useEffect(() => {
        
        setPuzzleGrid( [ cols, rows ] );
        
    }, [ setPuzzleGrid, cols, rows ]);

    return (
    
        <div id="GridInput">
            <p>Columns
                <NumberInput setValue={ setCols } min={ 4 } max={ 8 } initialValue={ puzzleGrid[0] } />
                {/* It send de previous context values to initial value of this component, to maintain them across all routes. */}
            </p>
            <p>Rows
                <NumberInput setValue={ setRows } min={ 3 } max={ 8 } initialValue={ puzzleGrid[1] } />
            </p>
        </div>
    );
}
