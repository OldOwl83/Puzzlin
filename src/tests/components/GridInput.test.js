import 'react';
import { shallow, mount } from 'enzyme';

import { GridInput } from "../../components/GridInput";

jest.mock( 'react', () => {
    return {
        ...jest.requireActual( 'react' ), 
        useContext: () => { return {
            puzzleGrid: [ 4, 3 ],
            setPuzzleGrid: jest.fn(),
        }}
    }
});

describe('Tests on the <GridInput /> component', () => {

    const setPuzzleGrid = jest.fn();

    test('It should match the snapshot.', () => {
      
        const wrapper = shallow( <GridInput /> );

        expect( wrapper ).toMatchSnapshot();
    });

    test('Its input boxes should take the context values as current values.', () => {
      
        const wrapper = mount( <GridInput /> );

        expect( wrapper.find( 'input' ).at(0).prop( 'value')).toBe( 4 );
        expect( wrapper.find( 'input' ).at(1).prop( 'value')).toBe( 3 );
    });
});
