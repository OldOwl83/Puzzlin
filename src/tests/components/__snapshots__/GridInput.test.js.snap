// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on the <GridInput /> component It should match the snapshot. 1`] = `
<div
  id="GridInput"
>
  <p>
    Columns
    <Memo()
      initialValue={4}
      max={8}
      min={4}
      setValue={[Function]}
    />
  </p>
  <p>
    Rows
    <Memo()
      initialValue={3}
      max={8}
      min={3}
      setValue={[Function]}
    />
  </p>
</div>
`;
